lang/java/avro/src/main/java/org/apache/avro/file/ZstandardCodec.java

```diff
@@ -24,22 +24,36 @@
 import java.io.OutputStream;
 import java.nio.ByteBuffer;
 
-import org.apache.commons.compress.compressors.zstandard.ZstdCompressorInputStream;
-import org.apache.commons.compress.compressors.zstandard.ZstdCompressorOutputStream;
 import org.apache.commons.compress.utils.IOUtils;
 
+import com.github.luben.zstd.Zstd;
+import com.github.luben.zstd.ZstdInputStream;
+import com.github.luben.zstd.ZstdOutputStream;
+
```

Comment:
Do we have to move from commons.compress to using the luben streams directly?   This introduces a hard dependency (so no longer optional/provided) which is less than ideal.    1.18 of commons.compress does include the compressionLevel in the constructor, but there may some other issue with it.    If we HAVE to use the luben things directly, we'll need to either update the CodecFactory code to handle the exception or resort to using some reflection.

Reply:
Yes, commons.compress does not expose the compression level.  Maybe if a newer version does we could use that.  It did not at the merge-base I used for the commit  (I have a 1.8.x branch with Zstandard in it).

But what is the point of commons.compress here?  It is just a thin wrapper that as far as I can see serves no real value.  For both Zstandard and XZ it at best wraps some underlying exceptions in commons.compress exception types.  That sounds like Codec's job to me.  (for bzip2 it actually provides a full java implementation of BZip2).

It does add one more jar in the classpath that might get mismatched and cause someone a bit of dependency hell.

The abstraction users care about is avro.file.{Codec,CodecFactory}, and the format of the bytes inside each data block.  

(Aside, should the zstandard compression also write the checksum for the block?  We can add that, or even make it optional; it is set to false by commons.compress).

>  we'll need to either update the CodecFactory code to handle the exception or resort to using some reflection.

I'm confused.  The exception that a user gets if they don't have luben.zstd in the classpath is _identical_ whether we directly access it or go through it via commons.compress  (ClassNotFound).  The stack is a little shorter, and happens at a slightly different place (see my other comment).

Personally, I'd love to get rid of commons.compress from avro entirely (fewer dependencies, i've had more than my fair share of far-too-interesting jar-hell issues in the hadoop ecosystem over the years) .  It would be easy to do other than the BZip2 case.   So if we make XZ and ZStandard not require commons.compress, then it becomes a provided/optional dependency only needed if you want BZip2.  And ... only legacy users need that.  Anything written today should use ZStandard as it can reach similar compression levels with much lower CPU use, of if you need the _ultimate_ in compression ratio and aren't too worried about compression time, xz can do a bit better and has faster decompression speed than BZip2.